File: bindgen-tests/python/tests/records.py
from test_package.uniffi_bindgen_tests import *
import unittest

class TestRecords(unittest.TestCase):
    def test_simple_record(self):
        self.assertEqual(roundtrip_simple_rec(SimpleRec(a=42)), SimpleRec(a=42))

    def test_unit_record(self):
        self.assertEqual(UnitRec(), UnitRec())

    def test_complex_record(self):
        self.assertEqual(
          roundtrip_complex_rec(
            ComplexRec(
              field_u8=0,
              field_i8=-1,
              field_u16=2,
              field_i16=-3,
              field_u32=4,
              field_i32=-5,
              field_u64=6,
              field_i64=-7,
              field_f32=8.5,
              field_f64=9.5,
              field_string="test",
              field_rec=SimpleRec(a=42)
            )
          ), ComplexRec(
            field_u8=0,
            field_i8=-1,
            field_u16=2,
            field_i16=-3,
            field_u32=4,
            field_i32=-5,
            field_u64=6,
            field_i64=-7,
            field_f32=8.5,
            field_f64=9.5,
            field_string="test",
            field_rec=SimpleRec(a=42)
          )
        )

if __name__ == '__main__':
    unittest.main()

File: bindgen-tests/lib/src/records.rs
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

#[derive(uniffi::Record)]
pub struct SimpleRec {
    pub a: u8,
}

#[derive(uniffi::Record)]
pub struct RecWithDefault {
    #[uniffi(default = 42)]
    pub a: u8,
}

#[derive(uniffi::Record)]
pub struct ComplexRec {
    pub field_u8: u8,
    pub field_i8: i8,
    pub field_u16: u16,
    pub field_i16: i16,
    pub field_u32: u32,
    pub field_i32: i32,
    pub field_u64: u64,
    pub field_i64: i64,
    pub field_f32: f32,
    pub field_f64: f64,
    #[uniffi(default = "DefaultString")]
    pub field_string: String,
    pub field_rec: SimpleRec,
}

#[uniffi::export]
pub fn roundtrip_simple_rec(rec: SimpleRec) -> SimpleRec {
    rec
}

#[uniffi::export]
pub fn roundtrip_complex_rec(rec: ComplexRec) -> ComplexRec {
    rec
}
